libcudacxx/include/cuda/std/__iterator/iterator_traits.h

```diff
@@ -165,38 +170,77 @@ struct __iter_traits_cache
-};
-struct __iter_concept_category_test
-{
-  template <class _Iter>
-  using _Apply = typename _ITER_TRAITS<_Iter>::iterator_category;
-};
-struct __iter_concept_random_fallback
-{
-  template <class _Iter>
-  using _Apply = enable_if_t<__is_primary_template<iterator_traits<_Iter>>::value, random_access_iterator_tag>;
-};
+#if _CCCL_STD_VER >= 2014
+#  if defined(_GLIBCXX_DEBUG)
+_CCCL_TEMPLATE(class _Iter, class _Ty, class _Range)
+_CCCL_REQUIRES(_IsSame<_Iter, ::__gnu_debug::_Safe_iterator<_Ty*, _Range>>::value)
+_LIBCUDACXX_HIDE_FROM_ABI auto
+  __iter_concept_fn(::__gnu_debug::_Safe_iterator<_Ty*, _Range>, __priority_tag<3>) -> contiguous_iterator_tag;
+#  endif
+#  if defined(__GLIBCXX__)
+_CCCL_TEMPLATE(class _Iter, class _Ty, class _Range)
+_CCCL_REQUIRES(_IsSame<_Iter, ::__gnu_cxx::__normal_iterator<_Ty*, _Range>>::value)
+_LIBCUDACXX_HIDE_FROM_ABI auto
+  __iter_concept_fn(::__gnu_cxx::__normal_iterator<_Ty*, _Range>, __priority_tag<3>) -> contiguous_iterator_tag;
+#  elif defined(_LIBCPP_VERSION)
+_CCCL_TEMPLATE(class _Iter, class _Ty)
+_CCCL_REQUIRES(_IsSame<_Iter, ::std::__wrap_iter<_Ty*>>::value)
+_LIBCUDACXX_HIDE_FROM_ABI auto __iter_concept_fn(::std::__wrap_iter<_Ty*>, __priority_tag<3>) -> contiguous_iterator_tag;
+#  elif defined(_MSVC_STL_VERSION) || defined(_IS_WRS)
+_CCCL_TEMPLATE(class _Iter)
+_CCCL_REQUIRES(_IsSame<_Iter, class _Iter::_Array_iterator>::value)
+_LIBCUDACXX_HIDE_FROM_ABI auto __iter_concept_fn(_Iter, __priority_tag<3>) -> contiguous_iterator_tag;
+_CCCL_TEMPLATE(class _Iter)
+_CCCL_REQUIRES(_IsSame<_Iter, class _Iter::_Array_const_iterator>::value)
+_LIBCUDACXX_HIDE_FROM_ABI auto __iter_concept_fn(_Iter, __priority_tag<3>) -> contiguous_iterator_tag;
+_CCCL_TEMPLATE(class _Iter)
+_CCCL_REQUIRES(_IsSame<_Iter, class _Iter::_Vector_iterator>::value)
+_LIBCUDACXX_HIDE_FROM_ABI auto __iter_concept_fn(_Iter, __priority_tag<3>) -> contiguous_iterator_tag;
+_CCCL_TEMPLATE(class _Iter)
+_CCCL_REQUIRES(_IsSame<_Iter, class _Iter::_Vector_const_iterator>::value)
+_LIBCUDACXX_HIDE_FROM_ABI auto __iter_concept_fn(_Iter, __priority_tag<3>) -> contiguous_iterator_tag;
+_CCCL_TEMPLATE(class _Iter)
+_CCCL_REQUIRES(_IsSame<_Iter, class _Iter::_String_iterator>::value)
+_LIBCUDACXX_HIDE_FROM_ABI auto __iter_concept_fn(_Iter, __priority_tag<3>) -> contiguous_iterator_tag;
+_CCCL_TEMPLATE(class _Iter)
+_CCCL_REQUIRES(_IsSame<_Iter, class _Iter::_String_const_iterator>::value)
+_LIBCUDACXX_HIDE_FROM_ABI auto __iter_concept_fn(_Iter, __priority_tag<3>) -> contiguous_iterator_tag;
+_CCCL_TEMPLATE(class _Iter)
+_CCCL_REQUIRES(_IsSame<_Iter, class _Iter::_String_view_iterator>::value)
+_LIBCUDACXX_HIDE_FROM_ABI auto __iter_concept_fn(_Iter, __priority_tag<3>) -> contiguous_iterator_tag;
+_CCCL_TEMPLATE(class _Iter)
+_CCCL_REQUIRES(_IsSame<_Iter, class _Iter::_Span_iterator>::value)
+_LIBCUDACXX_HIDE_FROM_ABI auto __iter_concept_fn(_Iter, __priority_tag<3>) -> contiguous_iterator_tag;
```

Comment:
You can probably do this more cleanly in modernish MSVC by detecting if our iterator unwrapping machinery produces a pointer from a given iterator type. `std::is_pointer_v<std::_Unwrapped_t<_Iter>>` is `true` for the STL's contiguous iterators, and for `gsl:span`'s iterators as well. (I didn't tell you to take this dependency on STL internals, I was never here. 😜)

Reply:
hrm. we need to support older versions of msvc. when did this change go in?

Reply:
I don't recall precisely when we defined the alias `_Unwrapped_t`, but `is_pointer_v<decltype(::std::_Get_unwrapped(::std::declval<_Iter>()))>` isn't much more to type, it should work back to at least VS2015. CE only goes back to 16.0 (https://www.godbolt.org/z/h4Wa1YvEc), and that works. Is that old enough?

Reply:
Works fine with VS2017 v15.9, not at all with VS2015.

Reply:
We only support 2017 gladly

Reply:
I had to revert the change to use `std::_Unwrapped_t`. something about that formulation was causing msvc to instantiate a template it shouldn't while doing overload resolution.